### ISSUE TITLE
UI infinite SSE creation + timeout on Chrome

### ISSUE THREAD

**Issue by gpamit** [argoproj/argo-workflows#12626]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

Argo WF GUI is not stable and not responsive. Most of the time its slow to load pages and running into error when click to show templates or workflows:
```
404 failed (timeout: https://[Link to ArgoWF]/monaco-editor.51a434cf513f802ff42b.js))
Stack Trace
ChunkLoadError: Loading chunk 404 failed.
....
```

This issue is frequent in Chrome, on the other hand Firefox is loading everything fine most of the time. 

Expected: Page should load without any error.

Versions:
```
Chrome: 121.0.6167.139 (Official Build) (x86_64)
Firefox: 122.0 (64-bit)
```

### Version

3.5.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
NA
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}

NA
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded


NA
```

**Comment by agilgur5** [comment]
> ```
> 404 failed (timeout: https://[Link to ArgoWF]/monaco-editor.51a434cf513f802ff42b.js))
> Stack Trace
> ChunkLoadError: Loading chunk 404 failed.
> ```

So this is specifically after the code-split of Monaco Editor in #12150. But if you're getting a timeout on that, you would've gotten a timeout on _every_ page before that PR.

> This issue is frequent in Chrome, on the other hand Firefox is loading everything fine most of the time.

This sounds like an issue with your device. What are the specs of your device? Is it low-end? How performant is your network connection? (download mbps)

I can't reproduce this and the UI has a lot of usage, so there isn't much actionable in the issue as is. If this is due to low end specs or poor network connection, continuing with #12059 will help with that, but unfortunately we can't split Monaco further than it already is. It's huge. The only other thing to do would be to replace it with a smaller and more efficient dep (which I did mention in the issue)

**Comment by gpamit** [comment]
@agilgur5 Thank you for looking into this. I can confirm this is problem for all my colleagues. 

Here is the system information

```
 System:
    OS: macOS 14.3
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 6.18 GB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Browsers:
    Chrome: 121.0.6167.139
    Safari: 17.3
```

I got decent download speed 150 MBPS. Since its happening with everyone in my company, it seems like a bug.

**Comment by agilgur5** [comment]
> ```
>     CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
>     Memory: 6.18 GB / 32.00 GB
> ```

> I got decent download speed 150 MBPS.

Huh, you're definitely on a performant machine and performant network connection, more performant than my own in both cases too, where I can't reproduce this.

> Since its happening with everyone in my company, it seems like a bug.

But it's not happening to all users (otherwise it would have been reported much earlier and have many +1s; also I would've noticed myself).

If it's everyone in your company, I wonder if it's a VPN or proxy issue? Or if it's the configuration of your `Ingress` in front of the Server? (for example, before I set-up SSO, I had used `oauth2-proxy` as a workaround, and that was a dramatic performance bottleneck)

**Comment by bradleyboveinis** [comment]
Myself and my colleagues are also experiencing this error.

Win 11 pro, 13900k, 128gb ram, 1gb down connection. Instance is accessed through teleport vpn. 

So far it only occurs for us on 1 instance, other team instances appear fine.

**Comment by agilgur5** [comment]
> started experiencing this error today.

The timeout? Did you do an update recently? What version are you on?
A lot more information and context would be helpful.

Again there isn't really anything actionable in the issue as-is. The one timeout error reported so far is for a dependency (Monaco) which is already as split as it can be, and so is no longer in Argo's control to shrink.

> So far it only occurs for us on 1 instance, other team instances appear fine.

Yes that would further suggest that a specific configuration is causing this and not a generic "UI is slow".
If that configuration is supposed to be performant but is actually slow, then a reproducible config environment would be the minimum necessary to debug anything or have something actionable.

**Comment by bradleyboveinis** [comment]
We've been on v3.5.2 for about 2 months. Our first report of the timeout issue came through today.  We're still working on trying to figure out what may be the cause as it's hard to reproduce. Previously it was consistently breaking in chrome, however it now appears to be occasionally working.

What I have observed, is that it only appears to happen in chrome for me, but works fine in FireFox. There appears to be stuck (pending) calls in chrome 

e.g.:
```
https://<uri>/api/v1/workflow-events/<ns>?listOptions.fieldSelector=metadata.namespace=mercury&listOptions.resourceVersion=1249507011&fields=result.object.metadata.name,result.object.metadata.namespace,result.object.metadata.resourceVersion,result.object.metadata.creationTimestamp,result.object.metadata.uid,result.object.status.finishedAt,result.object.status.phase,result.object.status.message,result.object.status.startedAt,result.object.status.estimatedDuration,result.object.status.progress,result.type,result.object.metadata.labels,result.object.metadata.annotations,result.object.spec.suspend
```
![image](https://github.com/argoproj/argo-workflows/assets/250562/8b40f77e-a2e5-46fc-a66a-e7615d99d521)

![image](https://github.com/argoproj/argo-workflows/assets/250562/62c366f2-cd4b-487a-bc05-57888f222aaf)

FF on the other hand looks to be using polling, whereas chrome appears to be trying to maintain a number of persistent connections.

At this point, I suspect it could be teleport vpn or another security feature preventing too many persistent connections from being opened.

**Comment by bradleyboveinis** [comment]
To replicate:

1. In chrome, open network tab
2. Navigate to `/workflows`
3. Wait until you have 6 persistent connections open. You should see something like this:
![image](https://github.com/argoproj/argo-workflows/assets/250562/eac90e90-6782-4916-8592-9698b56bb32a)
4. Try to navigate elsewhere

**Comment by stefansedich** [comment]
@agilgur5 I work with @bradleyboveinis and adding some more info here:

1. I ruled out Teleport by directly port-forwarding to the argo-server pod and still see the issue
2. I downgraded to 3.4.x and could not repliate the issue

On more caveat in our environment is we run with `--secure=false` which results in the server serving us http/1 which would result in the lower browser SSE connection limits and hence explain the issue we are seeing.

I know the UI changed between 3.4.x and 3.5.x to move to the unified workflows view, did this change how the SSEs are managed and how many may be used?

**Comment by agilgur5** [comment]
Update here, I seem to have found the main root cause in https://github.com/argoproj/argo-workflows/issues/12663#issuecomment-1947801740. That issue seems duplicative, or at least the root cause seems identical, but I'm not sure if the symptoms are exactly the same without more information from the user.
I'll have a fix out shortly that should be out in the next patch release.

Note that as I wrote there I was only able to partially reproduce this when enabling pagination and moving between pages. I was not able to reproduce this when doing nothing. I also only got two SSEs per page move, that would mostly add up on each other, but would never be more than two per page move. The root cause is likely the same, although I'm not sure how this infinite loop occurred as I couldn't repro it. 
I'm also not sure if this is the same as OP's issue -- network saturation or browser connection limits could cause any networking to slow down (or grind to a halt), but those wouldn't happen on the Workflow Details page/editing sidebar. But  yes, if they were on the Workflows List page first and then moved to those, the SSEs may have persisted (per the comment I linked, I couldn't figure out how they were persisting as the code specifically stops them and I confirmed it was running without errors).

Big thanks for all the details you both added here @bradleyboveinis and @stefansedich , as well as @alelapi in #12663! Those were all vital to partially reproducing it and figuring out what was going on 🙂

**Comment by agilgur5** [comment]
Responding to some questions & comments below:

> It appears as though there is a [hard limit of 6 connections per host in chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=12066)

Yea, that's a super rare limit I've occasionally hit into, usually only when you have an app that works with many tabs. There should only be one or two open connections for the Argo UI, so that was surprising to see. Makes sense based on the rest though -- nice job noticing that!

> 2. Log into argo and navigate to `/workflows`
> 3. Wait until you have 6 persistent connections open. You should see something like this:

This I could not fully reproduce, as I was only able to get two open at a time and only when moving between pages. But some old ones would remain after moving pages, causing a very similar effect at the end. 

> On more caveat in our environment is we run with `--secure=false` which results in the server serving us http/1 which would result in the lower browser SSE connection limits and potentially explain the issue we are seeing.

The SSEs not cancelling for some reason seems to be latency sensitive, so your usage might also have made the non-cancellation happen more frequently.

> I know the UI changed between 3.4.x and 3.5.x to move to the unified workflows view, did this change how the SSEs are managed and how many may be used when viewing workflows list?

It did not, but I landed a large refactor in 3.5.x as well in #11891. That actually significantly reduced the network activity as there were many unnecessary / duplicative requests being made before I refactored it (e.g. many list requests even though there's a `ListWatch` already, new lists and `ListWatch`es when no parts of the request changed, etc).
Unfortunately, I've also found 3 less than one-liner bugs in that refactor, including this one. One was a typo (https://github.com/argoproj/argo-workflows/issues/12663#issuecomment-1942137704), and then this one and the other one (#12562) were super nuanced, both being ref issues (recursive ref here, stale ref there -- that one also had historical codebase context/non-React usage that I didn't know about).

**Comment by agilgur5** [comment]
Regarding root causing and reproducing the infinite loop, I'm curious if there's maybe a stale cache or something causing that? That was something I was looking for more info on in https://github.com/argoproj/argo-workflows/issues/12663#issuecomment-1944425914. In local dev, I clear my caches with some frequency, so I wonder if that's why I haven't been able to repro it.

In particular would be `localStorage` that is used for initialization of the list filters, but cookies and other caches could have some effect too. For anyone here, does this still reproduce after clearing all caches?
